src/libraries/System.Private.CoreLib/src/System/IO/RandomAccess.Windows.cs

```diff
@@ -0,0 +1,566 @@
+        {
+            SafeFileHandle.ValueTaskSource vts = handle.GetValueTaskSource();
+            try
+            {
+                NativeOverlapped* nativeOverlapped = vts.PrepareForOperation(ReadOnlyMemory<byte>.Empty, fileOffset);
+                Debug.Assert(vts._memoryHandle.Pointer != null);
+
+                // Queue an async WriteFile operation.
+                if (Interop.Kernel32.WriteFileGather(handle, (Interop.Kernel32.FILE_SEGMENT_ELEMENT*)pinnedSegments.Pointer, bytesToWrite, IntPtr.Zero, nativeOverlapped) == 0)
+                {
+                    // The operation failed, or it's pending.
+                    int errorCode = FileStreamHelpers.GetLastWin32ErrorAndDisposeHandleIfInvalid(handle);
+                    if (errorCode == Interop.Errors.ERROR_IO_PENDING)
+                    {
+                        // Common case: IO was initiated, completion will be handled by callback.
+                        // Register for cancellation now that the operation has been initiated.
+                        vts.RegisterForCancellation(cancellationToken);
+                    }
+                    else
+                    {
+                        // Error. Callback will not be invoked.
+                        vts.Dispose();
+                        return errorCode == Interop.Errors.ERROR_NO_DATA // EOF on a pipe. IO callback will not be called.
+                            ? ValueTask.FromResult<int>(0)
+                            : ValueTask.FromException<int>(SafeFileHandle.ValueTaskSource.GetIOError(errorCode, path: null));
+                    }
+                }
+            }
+            catch
+            {
+                vts.Dispose();
+                throw;
+            }
+
+            // Completion handled by callback.
+            vts.FinishedScheduling();
+            return new ValueTask<int>(vts, vts.Version);
+        }
+
+        private static unsafe (MemoryHandle pinnedSegments, MemoryHandle[] memoryHandles, int totalBytes) PrepareMemorySegments(IReadOnlyList<Memory<byte>> buffers)
+        {
+            // "The array must contain enough elements to store nNumberOfBytesToWrite bytes of data, and one element for the terminating NULL. "
+            Interop.Kernel32.FILE_SEGMENT_ELEMENT[] fileSegments = new Interop.Kernel32.FILE_SEGMENT_ELEMENT[buffers.Count + 1];
+            fileSegments[buffers.Count].Buffer = IntPtr.Zero;
+
+            MemoryHandle[] memoryHandles = new MemoryHandle[buffers.Count];
+            int totalBytes = 0;
+            for (int i = 0; i < buffers.Count; i++)
+            {
+                Memory<byte> buffer = buffers[i];
+                MemoryHandle memoryHandle = buffer.Pin();
+                memoryHandles[i] = memoryHandle;
+                fileSegments[i] = new Interop.Kernel32.FILE_SEGMENT_ELEMENT { Buffer = new IntPtr(memoryHandle.Pointer) };
+
+                checked
+                {
+                    totalBytes += buffer.Length;
+                }
```

Comment:
This can throw.  If it does, you'll end up leaking the GCHandles on all memory handles already pinned.

Reply:
It should not throw as this method is called only for buffers for which the total length is less than `int.MaxSize`.

Reply:
I've modified it and I am just passing the pre-calculated size from previous method